### PR TITLE
feat(precondition-header): add new `precondition-header` spectral-style rule

### PR DESCRIPTION
**Pull request by pyrooka**
## PR summary

This commit introduces the new spectral-style
'precondition-header' rule. This rule checks operations
with `412` response defined, to make sure it supports
conditional header.

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

### General checklist
Please make sure that your PR fulfills the following requirements:  
- [x] The commit message follows the [Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Dependencies have been updated as needed
- [x] .secrets.baseline updated as needed?

#### Checklist for adding a new validation rule:
- [x] Added new validation rule definition (packages/ruleset/src/rules/*.js, index.js)
- [x] If necessary, added new validation rule implementation (packages/ruleset/src/functions/*.js, updated index.js)
- [x] Added new rule to default configuration (packages/ruleset/src/ibm-oas.js)
- [x] Added tests for new rule (packages/ruleset/test/*.test.js)
- [x] Added docs for new rule (docs/ibm-cloud-rules.md)

#### Checklist for removing an old validation rule:
- [ ] Removed old rule implementation (packages/validator/src/plugins/validation/*.js)
- [ ] Removed or adjusted testcases (packages/validator/test)
- [ ] Updated default configuration to deprecate old rule (packages/validator/src/.defaultsForValidator.js)
- [ ] Removed docs of old rule (docs/ibm-legacy-rules.md)

### CODE REVIEW

File: packages/ruleset/src/functions/precindition-header.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+module.exports = function(operation, _opts, context) {
```
Comment (padamstx) [inline packages/ruleset/src/functions/precindition-header.js:1]: please rename file to be `precondition-header.js`

**Review by padamstx — COMMENTED** [review]
Other than a few suggestions, this looks good.

However, I can't help but notice that the logic of this new rule seems to fit within the `response-status-codes` jurisdiction so to speak, so I think we should consider folding this logic into that rule.
However, if there is a need to keep this new rule's logic completely separate so that it can be independently enabled/disabled, then we can leave this new rule by itself, but I don't think that is the case.
@dpopp07 @hudlow what do you guyes think?

File: docs/ibm-cloud-rules.md [hunk 2]
```diff
@@ -341,6 +342,12 @@ or <code>application/merge-patch+json</code>.</td>
 <td>oas2, oas3</td>
 </tr>
 <tr>
+<td><a href="#rule-precondition-header">precondition-header</a></td>
+<td>error</td>
+<td>An operation with 412 response must support at least one conditional header</td>
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply docs/ibm-cloud-rules.md]: done

File: packages/ruleset/src/functions/precindition-header.js (hunk 1 above)
Comment (pyrooka) [reply packages/ruleset/src/functions/precindition-header.js]: Oops, done!

File: docs/ibm-cloud-rules.md [hunk 3]
```diff
@@ -3148,6 +3155,79 @@ paths:
...
 
+### Rule: precondition-header
+<table>
+<tr>
+<td><b>Rule id:</b></td>
+<td><b>precondition-header</b></td>
+</tr>
+<tr>
+<td valign=top><b>Description:</b></td>
+<td>An operation with 412 response must support at least one conditional header.</td>
```
[the remark this answers is not in the document]
Comment (pyrooka) [reply docs/ibm-cloud-rules.md]: done

**Comment by pyrooka** [comment]
@padamstx First off, thanks for the docs improvements! :)

Originally I started to implement this check in the `response-status-codes` rule, but that only has a warning severity while this one (from my understanding) should yield an error. Anyway, I'm open for any changes, let's discuss.

**Comment by padamstx** [comment]
> Originally I started to implement this check in the `response-status-codes` rule, but that only has warning severity while this one (from my understanding) should yield an error.

Yeah, that's true.  But to be honest, I don't know that it makes THAT much of a difference whether it's an error or warning.   Yes, it would mean that a violation of this guideline (if implemented as part of `response-status-codes`) would NOT cause a "bad result" when running the generator if it returns a warning instead of an error.  But users could modify the severity of the `response-status-codes` rule if it was that important.

My opinion (fwiw) is that the benefit of combining it all in `response-status-codes` slightly outweighs the negative of making it a warning.   @hudlow what's your take on this?

**Comment by hudlow** [comment]
This may not be a popular opinion, but I do not like big rules with generic categories of problems. In some cases (like pagination style) the benefits may be really significant (because of the implementation and performance efficiencies of doing various heuristics once for an operation versus every time), but in my view, this is a very neatly scoped rule, and `response-status-codes` may actually merit being broken up.

I generally don't like the idea of increasing the friction for a user to turn off rules in a granular way (often because of some legacy compatibility considerations for their API). If we have big rules with many requirements, a user is forced to either write individual overrides or turn off a bunch of other enforcement that they may have no other reason not to leave on.

But maybe I'm missing something that would make having hundreds of rules versus dozens (the direction I'm proposing we go) burdensome for us or for our users?

In any case, it seems like we should try to establish some sort of general policy for making this decision rather than it being arbitrary per enforcement point.

**Comment by dpopp07** [comment]
I'm inclined to agree with @hudlow on that point, I think I've seen the benefit of more granular customization for users even though it's more files to maintain. More files can mean some easier maintenance of individual bits of code though. And more rules doesn't mean more clutter from the users perspective, they just see the rules that pop up in their output and decide if they want to turn them off or not.

So for this rule, I think we can leave it as its own, configurable rule.

**Review by dpopp07 — APPROVED** [review]
Looks good! 👍

**Review by padamstx — APPROVED** [review]
LGTM

**Pull request merged**